### ISSUE TITLE
Larger batch size cause much more time in translation

### ISSUE THREAD

**Issue by IdiosyncraticDragon** [OpenNMT/OpenNMT-py#490]
I tried to use batch translation for a trained model (two layer Bahdanau netowork, WMT 14 EN-DE) on newstest2014. However it costed more time when the batch size increased.
When I used batch size 1 => it cost 335 seconds;
When I used batch size 100 => it cost 605 seconds;
When I used batch size 512 => it cost 815 seconds;
When I used batch size 1024 => it cost 786 seconds.

Is there any way to improve this? Because I think the purpose of batch model should be acceleration.

**Comment by srush** [comment]
Are you on CPU or GPU. Are you talking about time per batch or per sentence?

**Comment by IdiosyncraticDragon** [comment]
Translations  are run on P100, and the time is the total time including model loading and translation. 
- This is the part of calculating time:
```
starttime=`date +'%Y-%m-%d %H:%M:%S'`

CUDA_VISIBLE_DEVICES=5 ./my_bahdanau_issue8.sh $1

endtime=`date +'%Y-%m-%d %H:%M:%S'`
start_seconds=$(date --date="$starttime" +%s);
end_seconds=$(date --date="$endtime" +%s);
echo "Time： "$((end_seconds-start_seconds))"s"
```

- This is the part of translation (the file **my_bahdanau_issue8.sh**):
```
python ../translate.py -model /home/lgy/deepModels/torch_models/opennmt-py/$1 -src /home/lgy/data/wmt/wmt14-de-en/newstest2014-deen-src.en.tok -replace_unk -verbose -output pred_ende1.txt -gpu 0 -batch_size 512

perl ../tools/multi-bleu.perl /home/lgy/data/wmt/wmt14-de-en/newstest2014-deen-src.de.tok < pred_ende1.txt
```
- One thing I forgot to mention in description was that the model was trained previously which was based on branch 181fa18, not the newest version. Is this the problem?

**Comment by JianyuZhan** [comment]
You should make a clean run of train and translate on the same code base first to see if this problem still occurs.

**Comment by srush** [comment]
Thanks for the report though, we will try to replicate.

**Comment by IdiosyncraticDragon** [comment]
Well. I have trained two networks from scratch based on the newest code (branch 1695edca).  The eviroment has two 1080Ti and two Intel cpus with 2.1GHz. Here is the settings of trainig:
1. Dataset:
```
python ../preprocess.py -train_src /home/data/wmt/wmt14-de-en/from_sru/en-de/en-train.txt -train_tgt /home/data/wmt/wmt14-de-en/from_sru/en-de/de-train.txt -valid_src /home/data/wmt/wmt14-de-en/from_sru/en-de/en-val.txt -valid_tgt /home/data/wmt/wmt14-de-en/from_sru/en-de/de-val.txt -save_data /home/data/wmt/wmt14-de-en/len50_pywmt14 -src_vocab_size 50000 -tgt_vocab_size 50000 -src_seq_length 50 -tgt_seq_length 50 -dynamic_dict -max_shard_size 1000
``` 
1. Network 1=> SRUNet;  I used the checkpoint saved at epoch 12:
```
python ../train.py -data /home/data/wmt/wmt14-de-en/len50_pywmt14 -save_model /home/deepModels/torch_models/opennmt-py/sru -model_type 'text' -word_vec_size 500 -encoder_type 'rnn'  -rnn_size 500 -rnn_type 'SRU' -global_attention 'general' -layers 10 -input_feed 0 -optim 'sgd' -epochs 15 -batch_size 64 -dropout 0.1 -learning_rate 1.0 -learning_rate_decay 0.000001 -gpuid 0
```
2. Network 2=> Bahanau network; I used the checkpoint saved at epoch 2:
```
python ../train.py -data /home/data/wmt/wmt14-de-en/len50_pywmt14 -save_model /home/deepModels/torch_models/opennmt-py/deen_bahdanau -word_vec_size 620 -encoder_type 'brnn' -brnn_merge 'concat' -rnn_size 1000 -rnn_type 'GRU' -global_attention 'mlp' -optim 'adadelta' -batch_size 128 -epochs 20 -max_grad_norm 1 -gpuid 0 
```

Here is the results of translation (The trainslation script is similar to that I have mentioned in previous answers, and the time is the total time including model loading and traslation):
1. SRUNet (12 epochs):

|  Batch size |  Time (seconds)       | BLEU  |
| ------------- |:-------------:| -----:|
| 1                  | 248      | 17.03 |
| 100              | 440      |   2.04 |
| 512              | 435      |    0.78 |
| 1024            |  449     |    0.42 |

2. Bahanau (2 epochs, no results for batch size 1024 according to the memory limit)

|  Batch size |  Time (seconds)       | BLEU  |
| ------------- |:-------------:| -----:|
| 1                  | 251              | 14.34 |
| 100              | 518              |   14.34 |
| 512              | 548              |    14.34 |

It seems like the batch translation still has problem:
1. Its performance was not consist for SRU
2. All the time was increasing while the size of batch size increasing.

**Comment by guillaumekln** [comment]
Did you try with batch sizes between 1 and 100? And with a beam size of 1? 

The current beam search implementation is not exactly batch friendly and should certainly be rewritten for correctness and performance.

**Comment by IdiosyncraticDragon** [comment]
@guillaumekln As you said, the beam size really matters. Here is the results I just tested:

1. SRU (same settings as mentioned above)

Batch size | Beam size| Time (seconds) | BLEU
--      | --              | --   | --
1       |  5(default)| 248 | 17.03
100   |  5(default)| 440 | 2.04
512   |  5(default)| 435 | 0.78
1024 |  5(default)| 449 | 0.42
1       |  1             | 205 | 15.29
50     |  1             | 169 | 2.13
100   |  1             | 160 | 1.35
512   |  1             | 161| 0.47

2. Bahdanau (same settings as mentioned above)

Batch size | Beam size| Time (seconds) | BLEU
--    | --             | --   | --
1     |  5(default)| 251| 14.34
10   |  5(default)| 302 | 14.34
40   |  5(default)| 433 | 14.34
70   |  5(default)| 472 | 14.34
100 |  5(default)| 518 | 14.34
512 |  5(default)| 548 | 14.34
1     |  1| 200| 10.57
50   |  1| 173 | 10.57
100 |  1| 180 | 10.57
512 |  1| 165 | 10.57

I don't know whether there are other factors related, because even the beam size is only 1, the speed up of increasing batch size is not very significant.

**Comment by guillaumekln** [comment]
I think it the main cause is that finished hypotheses are not removed from the batch.

---

For reference, here are some numbers (tokens/second) for OpenNMT-Lua (see the "translate.lua (GPU)" column):

http://forum.opennmt.net/t/is-the-speed-of-ctranslate-faster-than-that-of-lua-version/224/11

**Comment by IdiosyncraticDragon** [comment]
The criteria of **tokens per second** seems better, I will rerun experiments and use the criteria to verify the differences.

**Comment by IdiosyncraticDragon** [comment]
Update (use **tokens per second** as metric instead of directly using **the total time**):

I insert such codes in translate.py (~ line 77) :
```
    import time # add
    start_t = time.time() # add

    for batch in test_data:
        .....
    end_t = time.time() # add
    total_sec = (end_t - start_t) # add
    print("%.4f tokens/sec" % (pred_words_total/total_sec)) # add
```
And rerun the experiments to get results below:

_**token per second**: the larger ones are better_

1. SRU

Batch size | Beam size | tokens/sec | BLEU
-- | -- | -- | --
1 | 5(default) | 271.82 | 17.03
10 | 5(default) | 142.42 | 7.01
40 | 5(default) | 91.29 | 3.45
70 | 5(default) | 90.70 | 2.75
100 | 5(default) | 91.92 | 2.04
512 | 5(default) | 100.39 | 0.78
1024 | 5(default) | 101.07 | 0.42
1 | 1 | 359.09 | 15.29
50 | 1 | 539.21 | 2.13
100 | 1 | 630.09 | 1.35
512 | 1 | 700.31 | 0.47

2. Bahdanau

Batch size | Beam size | tokens/sec | BLEU
-- | -- | -- | --
1 | 5(default) | 235.38 | 14.34
10 | 5(default) | 191.77 | 14.34
40 | 5(default) | 134.44 | 14.34
70 | 5(default) | 114.51 | 14.34
100 | 5(default) | 113.97 | 14.34
512 | 5(default) | 99.25 | 14.34
1 | 1 | 401.64 | 10.57
50 | 1 | 451.44 | 10.57
100 | 1 | 465.09 | 10.57
512 | 1 | 488.11 | 10.57
